raiden/waiting.py

```diff
@@ -369,38 +375,68 @@ def wait_for_healthy(
     wait_for_network_state(raiden, node_address, NODE_NETWORK_REACHABLE, retry_timeout)
 
 
-def wait_for_transfer_success(
+class TransferWaitResult(Enum):
+    SECRET_REGISTERED_ONCHAIN = "secret registered onchain"
+    UNLOCKED = "unlocked"
+    UNLOCK_FAILED = "unlock_failed"
+
+
+def wait_for_received_transfer_result(
     raiden: "RaidenService",
     payment_identifier: PaymentID,
     amount: PaymentAmount,
     retry_timeout: float,
-) -> None:  # pragma: no unittest
-    """Wait until a transfer with a specific identifier and amount
-    is seen in the WAL.
+    secrethash: SecretHash,
+) -> TransferWaitResult:  # pragma: no unittest
+    """Wait for the result of a transfer with the specified identifier
+    and/or secrethash. Possible results are onchain secret registration,
+    successful unlock and failed unlock. For a successful unlock, the
+    amount is also checked.
 
     Note:
         This does not time out, use gevent.Timeout.
     """
     log_details = {"payment_identifier": payment_identifier, "amount": amount}
-    found = False
-    while not found:
+    result = None
```

Comment:
This function only works with the receiving end of a transfer, so we don't care about the initiator here. For the mediators and the target we can recover the lock from the state changes, currently there is no function to load that from the database, you can use this as a template:

https://github.com/raiden-network/raiden/blob/ceabb9e1b628bd12129f67892c59c1e7ee6b4a50/raiden/storage/restore.py#L97-L129

I would name it as `get_state_change_by_lock_secrethash`, return the state change and then it can be used to get the lock expiration.